### PR TITLE
add h3d head

### PR DESCRIPTION
**Pull request by encore-zhou**

### CODE REVIEW

File: configs/h3dnet/h3dnet_8x8_scannet-3d-18class.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+                        [1.3766412, 0.65521795, 1.6813129],
+                        [0.6650819, 0.71111923, 1.298853],
+                        [0.41999173, 0.37906948, 1.7513971],
+                        [0.59359556, 0.5912492, 0.73919016],
+                        [0.50867593, 0.50656086, 0.30136237],
+                        [1.1511526, 1.0546296, 0.49706793],
+                        [0.47535285, 0.49249494, 0.5802117]])),
+    roi_head=dict(
+        primitive_list=[primitive_z_cfg, primitive_xy_cfg, primitive_line_cfg],
+        bbox_head=dict(
```
Comment (ZwwWayne) [inline configs/h3dnet/h3dnet_8x8_scannet-3d-18class.py:207]: What is the type of bbox_head?

File: configs/_base_/models/h3dnet.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        size_class_loss=dict(
+            type='CrossEntropyLoss', reduction='sum', loss_weight=1.0),
+        size_res_loss=dict(
+            type='SmoothL1Loss', reduction='sum', loss_weight=10.0),
+        semantic_loss=dict(
+            type='CrossEntropyLoss', reduction='sum', loss_weight=1.0)),
+    roi_head=dict(
+        type='H3DRoIHead',
+        bbox_head=dict(
+            type='H3dHead',
```
Comment (ZwwWayne) [inline configs/_base_/models/h3dnet.py:114]: H3dHead -> H3DHead

File: configs/h3dnet/h3dnet_8x8_scannet-3d-18class.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        var_thresh=1e-2,
+        lower_thresh=1e-6,
+        num_point=100,
+        num_point_line=10,
+        line_thresh=0.2))
+
+# model settings
+model = dict(
+    backbone=dict(
+        type='MultiBackbone',
```
Comment (ZwwWayne) [inline configs/h3dnet/h3dnet_8x8_scannet-3d-18class.py:162]: Please try to reduce settings in this config that duplicates the base config.

File: mmdet3d/models/detectors/h3dnet.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+            points (list[torch.Tensor]): Points of each sample.
+            img_metas (list): Image metas.
+            rescale (bool): Whether to rescale results.
+
+        Returns:
+            list: Predicted 3d boxes.
+        """
+        points_cat = torch.stack(points)
+
+        x = self.extract_feat(points_cat)
```
Comment (xavierwu95) [inline mmdet3d/models/detectors/h3dnet.py:108]: modify as in self.forward_train

File: mmdet3d/models/detectors/h3dnet.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+        # after merging, bboxes will be rescaled to the original image size
+        merged_bboxes = merge_aug_bboxes_3d(aug_bboxes, img_metas,
+                                            self.bbox_head.test_cfg)
+
+        return merged_bboxes
+
+    def extract_feats(self, points, img_metas):
+        """Extract features of multiple samples."""
+        return [
```
Comment (xavierwu95) [inline mmdet3d/models/detectors/h3dnet.py:168]: 1. docstrings
2. it seems like `points` is a Tensor, so is it necessary to use [... for ... in ...] ?

File: mmdet3d/models/roi_heads/bbox_heads/h3d_head.py [hunk 6]
```diff
@@ -0,0 +1,1017 @@
...
+                loss for each prediction bounding box.
+            box_loss_weights (torch.Tensor): Weights of regression \
+                loss for each prediction bounding box.
+            valid_gt_weights (torch.Tensor): Validation of each \
+                ground truth bounding box.
+
+        Returns:
+            dict: Losses of aggregation module.
+        """
+        device = bbox_preds['dir_res_norm' + suffix].device
```
Comment (xavierwu95) [inline mmdet3d/models/roi_heads/bbox_heads/h3d_head.py:324]: do not use .device. use a.new_zeros() instead

File: mmdet3d/models/roi_heads/bbox_heads/h3d_head.py [hunk 7]
```diff
@@ -0,0 +1,1017 @@
...
+        # calculate direction class loss
+        dir_class_loss = self.dir_class_loss(
+            bbox_preds['dir_class' + suffix].transpose(2, 1),
+            dir_class_targets,
+            weight=box_loss_weights)
+
+        # calculate direction residual loss
+        batch_size, proposal_num = size_class_targets.shape[:2]
+        heading_label_one_hot = torch.zeros(
+            (batch_size, proposal_num, self.num_dir_bins)).to(device)
```
Comment (xavierwu95) [inline mmdet3d/models/roi_heads/bbox_heads/h3d_head.py:348]: use a.new_zeros() instead

File: mmdet3d/models/roi_heads/bbox_heads/h3d_head.py [hunk 8]
```diff
@@ -0,0 +1,1017 @@
...
+
+        # calculate size class loss
+        size_class_loss = self.size_class_loss(
+            bbox_preds['size_class' + suffix].transpose(2, 1),
+            size_class_targets,
+            weight=box_loss_weights)
+
+        # calculate size residual loss
+        one_hot_size_targets = torch.zeros(
+            (batch_size, proposal_num, self.num_sizes)).to(device)
```
Comment (xavierwu95) [inline mmdet3d/models/roi_heads/bbox_heads/h3d_head.py:363]: use a.new_zeros() instead

File: mmdet3d/models/roi_heads/bbox_heads/h3d_head.py [hunk 9]
```diff
@@ -0,0 +1,1017 @@
...
+            dir_res_targets,
+            center_targets,
+            mask_targets,
+            objectness_targets,
+            objectness_weights,
+            box_loss_weights,
+            valid_gt_weights,
+            suffix='_opt')
+        for key in refined_proposal_loss.keys():
+            losses[key + '_opt'] = refined_proposal_loss[key]
```
Comment (xavierwu95) [inline mmdet3d/models/roi_heads/bbox_heads/h3d_head.py:170]: use the full name of `_opt`

File: mmdet3d/models/roi_heads/bbox_heads/h3d_head.py [hunk 10]
```diff
@@ -0,0 +1,1017 @@
...
+            prev_channel = primitive_refine_channels[k]
+        # Final object proposal/detection
+        # Objectness scores (2), center residual (3),
+        # heading class+residual (num_heading_bin*2), size class +
+        # residual(num_size_cluster*4)
+        conv_out_channel = (2 + 3 + num_heading_bin * 2 +
+                            num_size_cluster * 4 + self.num_classes)
+        self.proposal_pred.append(nn.Conv1d(prev_channel, conv_out_channel, 1))
+
+        self.softmax_normal = torch.nn.Softmax(dim=1)
```
Comment (xavierwu95) [inline mmdet3d/models/roi_heads/bbox_heads/h3d_head.py:549]: self.softmax_normal is not called.

File: mmdet3d/models/roi_heads/bbox_heads/h3d_head.py [hunk 11]
```diff
@@ -0,0 +1,1017 @@
...
+                of surface center.
+            pred_obj_line_center (torch.Tensor): Objectness prediction of \
+                line center.
+            pred_surface_sem (torch.Tensor): Semantic prediction of \
+                surface center.
+            pred_line_sem (torch.Tensor): Semantic prediction of line center.
+        Returns:
+            tuple[torch.Tensor]: Targets for primitive cues.
+        """
+        device = points.device
```
Comment (xavierwu95) [inline mmdet3d/models/roi_heads/bbox_heads/h3d_head.py:886]: do not use .device

File: mmdet3d/models/roi_heads/bbox_heads/h3d_head.py [hunk 12]
```diff
@@ -0,0 +1,1017 @@
...
+        gt_bboxes_3d = gt_bboxes_3d.to(device)
```
Comment (xavierwu95) [inline mmdet3d/models/roi_heads/bbox_heads/h3d_head.py:887]: is it necessary to set device here?

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 13]
```diff
@@ -0,0 +1,754 @@
...
+            gt_labels_3d (torch.Tensor): Labels of each batch.
+            pts_semantic_mask (None | torch.Tensor): Point-wise semantic
+                label of each batch.
+            pts_instance_mask (None | torch.Tensor): Point-wise instance
+                label of each batch.
+
+        Returns:
+            tuple[torch.Tensor]: Targets of primitive head.
+        """
+        gt_bboxes_3d = gt_bboxes_3d.to(points.device)
```
Comment (xavierwu95) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:350]: is it necessary to use `to(points.device)`?

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 14]
```diff
@@ -0,0 +1,754 @@
...
+        point_offset = points.new_zeros([num_points, 3])
+        point_sem = points.new_zeros([num_points, 3 + self.num_dim + 1])
+
+        instance_flag = torch.nonzero(
+            pts_semantic_mask != self.num_classes).squeeze(1)
+        instance_labels = pts_instance_mask[instance_flag].unique()
+
+        for i, i_instance in enumerate(instance_labels):
+            ind = instance_flag[pts_instance_mask[instance_flag] == i_instance]
+            x = points[ind, :3]
```
Comment (xavierwu95) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:363]: x -> coords

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 15]
```diff
@@ -0,0 +1,754 @@
...
+        point_mask = points.new_zeros(num_points)
+        point_offset = points.new_zeros([num_points, 3])
+        point_sem = points.new_zeros([num_points, 3 + self.num_dim + 1])
+
+        instance_flag = torch.nonzero(
+            pts_semantic_mask != self.num_classes).squeeze(1)
+        instance_labels = pts_instance_mask[instance_flag].unique()
+
+        for i, i_instance in enumerate(instance_labels):
+            ind = instance_flag[pts_instance_mask[instance_flag] == i_instance]
```
Comment (xavierwu95) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:362]: ind -> indices

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 16]
```diff
@@ -0,0 +1,754 @@
...
+                        center[0], center[1], center[2], xmax - xmin,
+                        ymax - ymin, (pts_semantic_mask[ind][0])
+                    ])
+                    point_offset[sel_global] = center - x[sel]
+
+            # Get the boundary points here
+            alldist = torch.abs(
+                torch.sum(x * plane_upper[:3], 1) + plane_upper[-1])
+            mind = alldist.min()
+            sel = torch.abs(alldist - mind) < self.train_cfg['dist_thresh']
```
Comment (xavierwu95) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:431]: use the full name of `sel`

File: mmdet3d/models/roi_heads/bbox_heads/h3d_head.py [hunk 17]
```diff
@@ -0,0 +1,1017 @@
...
+from mmdet3d.core.post_processing import aligned_3d_nms
+from mmdet3d.models.builder import build_loss
+from mmdet3d.models.losses import chamfer_distance
+from mmdet3d.ops import PointSAModule
+from mmdet.core import build_bbox_coder, multi_apply
+from mmdet.models import HEADS
+
+
+@HEADS.register_module()
+class H3dHead(nn.Module):
```
Comment (xavierwu95) [inline mmdet3d/models/roi_heads/bbox_heads/h3d_head.py:16]: H3DBboxHead
also change the file name

File: mmdet3d/models/roi_heads/bbox_heads/h3d_head.py [hunk 18]
```diff
@@ -0,0 +1,1017 @@
...
+         valid_gt_masks, objectness_targets, objectness_weights,
+         box_loss_weights, valid_gt_weights) = rpn_targets
+
+        loss_inputs = (bbox_preds, points, gt_bboxes_3d, gt_labels_3d,
+                       pts_semantic_mask, pts_instance_mask, img_metas,
+                       gt_bboxes_ignore)
+        losses = {}
+
+        # calculate refined proposal loss
+        refined_proposal_loss = self.get_proposal_stage_loss(
```
Comment (xavierwu95) [inline mmdet3d/models/roi_heads/bbox_heads/h3d_head.py:156]: what is the different between `self.get_proposal_stage_loss` and loss in VoteHead

Comment (encore-zhou) [reply mmdet3d/models/roi_heads/bbox_heads/h3d_head.py:156]: support paramaters `suffix`, inorder to compute bbox regression loss and classification loss for different stages.

File: mmdet3d/models/detectors/h3dnet.py [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+                label of each batch.
+            gt_bboxes_ignore (None | list[torch.Tensor]): Specify
+                which bounding.
+
+        Returns:
+            dict: Losses.
+        """
+        points_cat = torch.stack(points)
+
+        x = self.extract_feat(points_cat)
```
[the remark this answers is not in the document]
Comment (encore-zhou) [reply mmdet3d/models/detectors/h3dnet.py:59]: As we use `VoteHead` as the rpn stage and `VoteHead`  uses key `fp_xyz` to extract input points, so we need to rename the outputs dict of `MultiBackbone`

File: mmdet3d/models/roi_heads/bbox_heads/h3d_head.py (hunk 18 above)
Comment (encore-zhou) [reply mmdet3d/models/roi_heads/bbox_heads/h3d_head.py:156]: And also remove the loss of `vote_module`

File: mmdet3d/models/detectors/h3dnet.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (xavierwu95) [reply mmdet3d/models/detectors/h3dnet.py:59]: x -> feats_dict

File: mmdet3d/models/detectors/h3dnet.py (hunk 4 above)
Comment (xavierwu95) [inline mmdet3d/models/detectors/h3dnet.py:108]: x->feat_dict

File: mmdet3d/models/detectors/h3dnet.py [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+        x['fp_xyz'] = [x['fp_xyz_net0'][-1]]
+        x['fp_features'] = [x['hd_feature']]
+        x['fp_indices'] = [x['fp_indices_net0'][-1]]
+
+        if self.with_rpn:
+            rpn_outs = self.rpn_head(x, self.test_cfg.rpn.sample_mod)
+            x.update(rpn_outs)
+            # Generate rpn proposals
+            rpn_proposals = self.rpn_head.get_bboxes(
+                points, rpn_outs, img_metas, with_nms=False)
```
Comment (xavierwu95) [inline mmdet3d/models/detectors/h3dnet.py:118]: with_nms=False should be defined in test_cfg
rpn_proposals -> proposal_list

File: mmdet3d/models/detectors/h3dnet.py [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+            x['rpn_proposals'] = rpn_proposals
+        else:
+            raise NotImplementedError
+
+        return self.roi_head.simple_test(
+            x,
+            self.test_cfg.rcnn.sample_mod,
```
Comment (xavierwu95) [inline mmdet3d/models/detectors/h3dnet.py:125]: modify as in self.forward_train

File: mmdet3d/models/roi_heads/bbox_heads/h3d_bbox_head.py [hunk 22]
```diff
@@ -0,0 +1,1015 @@
...
+                    matching_feat_dims,
+                    1,
+                    padding=0,
+                    conv_cfg=conv_cfg,
+                    norm_cfg=norm_cfg,
+                    bias=True,
+                    inplace=True))
+        self.line_feat_aggregation = nn.Sequential(*self.line_feat_aggregation)
+
+        prev_channel = 18 * matching_feat_dims
```
Comment (xavierwu95) [inline mmdet3d/models/roi_heads/bbox_heads/h3d_bbox_head.py:527]: illustrate the magic number 18

File: mmdet3d/core/bbox/structures/depth_box3d.py [hunk 23]
```diff
@@ -251,3 +251,52 @@ def points_in_boxes(self, points):
         box_idxs_of_pts = points_in_boxes_batch(points_lidar, boxes_lidar)
 
         return box_idxs_of_pts.squeeze(0)
+
+    def get_surface_line_center(self):
```
Comment (xavierwu95) [inline mmdet3d/core/bbox/structures/depth_box3d.py:255]: add unittest for this function

File: mmdet3d/models/detectors/h3dnet.py [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+            proposal_cfg = self.train_cfg.get('rpn_proposal',
+                                              self.test_cfg.rpn)
+            proposal_list = self.rpn_head.get_bboxes(
+                points, rpn_outs, img_metas, use_nms=proposal_cfg.use_nms)
+            feats_dict['proposal_list'] = proposal_list
+        else:
+            raise NotImplementedError
+
+        roi_losses = self.roi_head.forward_train(
+            feats_dict, self.train_cfg.rcnn.sample_mod, img_metas, points,
```
Comment (xavierwu95) [inline mmdet3d/models/detectors/h3dnet.py:89]: the config of rcnn should be passed as [this](https://github.com/open-mmlab/mmdetection3d/blob/master/mmdet3d/models/roi_heads/part_aggregation_roi_head.py#L31)

File: mmdet3d/models/detectors/h3dnet.py [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+        feats_dict['fp_features'] = [feats_dict['hd_feature']]
+        feats_dict['fp_indices'] = [feats_dict['fp_indices_net0'][-1]]
+
+        if self.with_rpn:
+            proposal_cfg = self.test_cfg.rpn
+            rpn_outs = self.rpn_head(feats_dict, proposal_cfg.sample_mod)
+            feats_dict.update(rpn_outs)
+            # Generate rpn proposals
+            proposal_list = self.rpn_head.get_bboxes(
+                points, rpn_outs, img_metas, use_nms=proposal_cfg.use_nms)
```
Comment (xavierwu95) [inline mmdet3d/models/detectors/h3dnet.py:120]: modify as in self.forward_train

File: mmdet3d/models/detectors/h3dnet.py [hunk 26]
```diff
@@ -0,0 +1,174 @@
...
+            img_metas,
+            points_cat,
+            rescale=rescale)
+
+    def aug_test(self, points, img_metas, imgs=None, rescale=False):
+        """Test with augmentation."""
+        points_cat = [torch.stack(pts) for pts in points]
+        feats_dict = self.extract_feats(points_cat, img_metas)
+        for x in feats_dict:
+            x['fp_xyz'] = [x['fp_xyz_net0'][-1]]
```
Comment (xavierwu95) [inline mmdet3d/models/detectors/h3dnet.py:137]: x -> feats_dict

File: mmdet3d/models/detectors/h3dnet.py [hunk 27]
```diff
@@ -0,0 +1,174 @@
...
+        # only support aug_test for one sample
+        aug_bboxes = []
+        for x, pts_cat, img_meta in zip(feats_dict, points_cat, img_metas):
+            if self.with_rpn:
+                proposal_cfg = self.test_cfg.rpn
+                rpn_outs = self.rpn_head(x, proposal_cfg.sample_mod)
+                x.update(rpn_outs)
+                # Generate rpn proposals
+                proposal_list = self.rpn_head.get_bboxes(
+                    points, rpn_outs, img_metas, use_nms=proposal_cfg.use_nms)
```
Comment (xavierwu95) [inline mmdet3d/models/detectors/h3dnet.py:150]: modify as in self.forward_train

File: mmdet3d/models/roi_heads/h3d_roi_head.py [hunk 28]
```diff
@@ -0,0 +1,165 @@
...
+        bbox_head['test_cfg'] = self.test_cfg
+        self.bbox_head = build_head(bbox_head)
+
+    def init_assigner_sampler(self):
+        """Initialize assigner and sampler."""
+        pass
+
+    def forward_train(self,
+                      feats_dict,
+                      sample_mod,
```
Comment (xavierwu95) [inline mmdet3d/models/roi_heads/h3d_roi_head.py:53]: move `sample_mod` to config

File: mmdet3d/models/detectors/h3dnet.py [hunk 29]
```diff
@@ -0,0 +1,174 @@
...
+            rpn_outs = self.rpn_head(feats_dict, self.train_cfg.rpn.sample_mod)
+            feats_dict.update(rpn_outs)
+
+            rpn_loss_inputs = (points, gt_bboxes_3d, gt_labels_3d,
+                               pts_semantic_mask, pts_instance_mask, img_metas)
+            rpn_losses = self.rpn_head.loss(
+                rpn_outs,
+                *rpn_loss_inputs,
+                gt_bboxes_ignore=gt_bboxes_ignore,
+                ret_target=True)
```
[the remark this answers is not in the document]
Comment (encore-zhou) [reply mmdet3d/models/detectors/h3dnet.py:75]: `ret_target` is not a common attribute of `Votehead`

**Review by xavierwu95 — APPROVED** [review]

File: mmdet3d/models/dense_heads/vote_head.py [hunk 30]
```diff
@@ -491,14 +498,20 @@ def get_targets_single(self,
...
 
         Args:
             points (torch.Tensor): Input points.
             bbox_preds (dict): Predictions from vote head.
             input_metas (list[dict]): Point cloud and image's meta info.
             rescale (bool): Whether to rescale bboxes.
+            use_nms (bool): Whether to apply NMS.
 
         Returns:
             list[tuple[torch.Tensor]]: Bounding boxes, scores and labels.
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/vote_head.py:517]: Update docstring

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**